src/box/lua/config/utils/aboard.lua

```diff
@@ -0,0 +1,95 @@
+-- Alert board.
+
+local datetime = require('datetime')
+local log = require('internal.config.utils.log')
+
+local function aboard_set(self, key, alert)
+    assert(alert.type == 'error' or alert.type == 'warn')
+    if alert.type == 'error' then
+        log.error(alert.message)
+    else
+        log.warn(alert.message)
+    end
+    alert.timestamp = datetime.now()
+    self._alerts[key] = alert
+end
+
+local function aboard_get(self, key)
+    return self._alerts[key]
+end
+
+local function aboard_drop(self, key)
+    self._alerts[key] = nil
+    if self._on_drop ~= nil then
+        self._on_drop(self, key)
+    end
+end
+
+local function aboard_clean(self)
+    self._alerts = {}
+end
+
+local function aboard_alerts(self)
+    -- Don't return alert keys.
+    local alerts = {}
+    for _, alert in pairs(self._alerts) do
+        table.insert(alerts, alert)
+    end
+    -- Sort by timestamps.
+    table.sort(alerts, function(a, b)
+        return a.timestamp < b.timestamp
+    end)
+    return alerts
+end
+
+-- Return one of three possible statuses:
+--
+-- * ready
+-- * check_warnings
+-- * check_errors
+local function aboard_status(self)
+    local status = 'ready'
+
+    for _, alert in pairs(self._alerts) do
+        assert(alert.type == 'error' or alert.type == 'warn')
+        if alert.type == 'error' then
+            return 'check_errors'
+        end
+
+        status = 'check_warnings'
+    end
```

Comment:
It changes a meaning of the statuses. An instance reports `startup_in_progress` or `reload_in_progress` while it is in a transient state. The `ready`, `check_warnings` or `check_errors` status is set only after applying a configuration.

`aboard_status()` is called at end of the configuration applying. `config:info()` returns the saved value.

Reply:
Let's discuss it all together in the [thread above](https://github.com/tarantool/tarantool/pull/9598#discussion_r1457093321).

Reply:
OTOH, if it you mean aboard's status (not config's status) or some `has_warning`/`has_error` flags, it has sense.

I'll experiment around and update it this way if the code will not look too complicated for me.

Reply:
Well, yes, this crossed my mind, maybe having two state machines would help. But that needs to be looked into.

Reply:
I've experimented around, see below.

```diff
diff --git a/src/box/lua/config/utils/aboard.lua b/src/box/lua/config/utils/aboard.lua
index 480f4c98d2..9b572a93cb 100644
--- a/src/box/lua/config/utils/aboard.lua
+++ b/src/box/lua/config/utils/aboard.lua
@@ -28,6 +28,7 @@ local log = require('internal.config.utils.log')
 local function aboard_set(self, key, alert)
     assert(alert.type == 'error' or alert.type == 'warn')
     if alert.type == 'error' then
+        self._error_count = self._error_count + 1
         log.error(alert.message)
     else
         log.warn(alert.message)
@@ -48,9 +49,13 @@ end
 -- The `on_drop` callback (see the aboard.new() function) is
 -- called if the alert had exist.
 local function aboard_drop(self, key)
-    if self._alerts[key] == nil then
+    local alert = self._alerts[key]
+    if alert == nil then
         return
     end
+    if alert.type == 'error' then
+        self._error_count = self._error_count - 1
+    end
     self._alerts[key] = nil
     if self._on_drop ~= nil then
         self._on_drop(self, key)
@@ -61,6 +66,7 @@ end
 --
 -- The `on_drop` callback is NOT called.
 local function aboard_clean(self)
+    self._error_count = 0
     self._alerts = {}
 end
 
@@ -91,18 +97,13 @@ end
 -- * check_warnings
 -- * check_errors
 local function aboard_status(self)
-    local status = 'ready'
-
-    for _, alert in pairs(self._alerts) do
-        assert(alert.type == 'error' or alert.type == 'warn')
-        if alert.type == 'error' then
-            return 'check_errors'
-        end
-
-        status = 'check_warnings'
+    if self._error_count > 0 then
+        return 'check_errors'
+    elseif self:is_empty() then
+        return 'ready'
+    else
+        return 'check_warnings'
     end
-
-    return status
 end
 
 -- Return `true` if there are no alerts, `false` otherwise.
@@ -139,6 +140,7 @@ local function new(opts)
     end
 
     return setmetatable({
+        _error_count = 0,
         _alerts = {},
         _on_drop = on_drop,
     }, mt)
```

TBH, I don't like to tweak the counter in several methods. It looks as an unnecessary complication.

I expect that amount of alerts is small enough to just traverse them on an event.

I'll leave it as is now. Let's discuss.

Reply:
We've discussed it more privately and agreed that `status` is required only after the loading, so it is ok to scan the alerts afterwards.